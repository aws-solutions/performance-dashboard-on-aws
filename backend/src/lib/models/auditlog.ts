/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
