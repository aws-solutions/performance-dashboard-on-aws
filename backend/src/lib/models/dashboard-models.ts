export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};