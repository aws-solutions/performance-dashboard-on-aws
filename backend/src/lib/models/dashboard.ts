/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { Widget } from "./widget";

export const DASHBOARD_ITEM_TYPE = "Dashboard";

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    /**
     * @type {string}
     * @deprecated in favor of {@link Draft}
     */
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export interface DashboardVersion {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
}

export interface Dashboard {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: Date;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: DashboardState;
    releaseNotes?: string;
    widgets?: Array<Widget>;
    friendlyURL?: string;
}

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string;
    sk: string;
    type: string;
    version: number;
    parentDashboardId: string;
    dashboardName: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: string;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: string;
    releaseNotes?: string;
    friendlyURL?: string;
}

// Public representation of a dashboard. Hides some fields
// like createdBy and state, which should not be exposed
// on public endpoints.
export interface PublicDashboard {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    updatedAt: Date;
    widgets?: Array<Widget>;
    friendlyURL?: string;

    // Filled on a homepage search with dashboard content
    // that matches the search query.
    queryMatches?: Array<string>;
}
