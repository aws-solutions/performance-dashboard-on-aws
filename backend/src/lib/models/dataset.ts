/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export interface Dataset {
    id: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: Date;
    sourceType: SourceType;
    schema?: DatasetSchema;
}

export type DatasetContent = Array<object>;
export type DatasetList = Array<Dataset>;

export interface DatasetItem {
    pk: string;
    sk: string;
    type: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: string;
    sourceType: string;
    schema?: string;
}
