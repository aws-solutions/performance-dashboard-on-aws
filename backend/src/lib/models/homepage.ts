/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export interface Homepage {
    title: string;
    description: string;
    updatedAt: Date;
}

export interface HomepageItem {
    pk: string;
    sk: string;
    type: string;
    updatedAt?: string;
    title: string;
    description: string;
}
