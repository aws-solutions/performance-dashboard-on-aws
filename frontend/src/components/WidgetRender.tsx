/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import {
    Widget,
    WidgetType,
    ChartWidget,
    TableWidget,
    ImageWidget,
    MetricsWidget,
} from "../models";
import { useWidgetDataset, useImage } from "../hooks";
import ChartWidgetComponent from "../components/ChartWidget";
import TableWidgetComponent from "../components/TableWidget";
import TextWidget from "../components/TextWidget";
import SectionWidget from "../components/SectionWidget";
import ImageWidgetComponent from "../components/ImageWidget";
import MetricsWidgetComponent from "../components/MetricsWidget";
import Utils from "../services/UtilsService";
import styles from "./WidgetRender.module.scss";

interface Props {
    widget: Widget;
    showMobilePreview?: boolean;
    hideTitle?: boolean;
    widgets?: Array<Widget>;
    setActiveWidgetId?: Function;
    topOffset?: string;
    bottomOffset?: string;
    defaultActive?: string;
    disableShare?: boolean;
}

function WidgetRender({
    widget,
    showMobilePreview,
    widgets,
    hideTitle,
    setActiveWidgetId,
    bottomOffset,
    topOffset,
    defaultActive,
    disableShare,
}: Props) {
    function getWidget() {
        switch (widget.widgetType) {
            case WidgetType.Text:
                return <TextWidget widget={widget} hideTitle={hideTitle} />;
            case WidgetType.Chart:
                return (
                    <ChartWidgetComponent
                        widget={widget as ChartWidget}
                        showMobilePreview={showMobilePreview}
                        hideTitle={hideTitle}
                    />
                );
            case WidgetType.Table:
            case WidgetType.Metrics:
                return (
                    <WidgetWithDataset
                        widget={widget}
                        showMobilePreview={showMobilePreview}
                        hideTitle={hideTitle}
                    />
                );
            case WidgetType.Image:
                return <WidgetWithImage widget={widget as ImageWidget} hideTitle={hideTitle} />;
            case WidgetType.Section:
                return (
                    <SectionWidget
                        widget={widget}
                        showMobilePreview={showMobilePreview}
                        widgets={widgets}
                        setActiveWidgetId={setActiveWidgetId}
                        bottomOffset={bottomOffset}
                        topOffset={topOffset}
                        defaultActive={defaultActive}
                    />
                );
            default:
                return null;
        }
    }

    if (!!disableShare) {
        return getWidget();
    }

    return <div className={styles.widgetContainer}>{getWidget()}</div>;
}

function WidgetWithImage({ widget, hideTitle }: Props) {
    const imageWidget = widget as ImageWidget;
    const content = imageWidget.content;
    const file = useImage(content.s3Key.raw);

    const imageId = `image-${Utils.getShorterId(widget.id)}`;
    return (
        <ImageWidgetComponent
            id={imageId}
            title={!hideTitle && imageWidget.showTitle ? content.title : ""}
            summary={content.summary ? content.summary : ""}
            summaryBelow={content.summaryBelow}
            file={file.file}
            altText={content.imageAltText}
            scalePct={content.scalePct ? content.scalePct : "auto"}
        />
    );
}

function WidgetWithDataset({ widget, showMobilePreview, hideTitle }: Props) {
    const { json } = useWidgetDataset(widget);
    switch (widget.widgetType) {
        case WidgetType.Table:
            const tableWidget = widget as TableWidget;
            const tableId = `table-${Utils.getShorterId(widget.id)}`;
            return (
                <TableWidgetComponent
                    id={tableId}
                    title={!hideTitle && tableWidget.showTitle ? tableWidget.content.title : ""}
                    summary={tableWidget.content.summary}
                    data={json}
                    summaryBelow={tableWidget.content.summaryBelow}
                    columnsMetadata={tableWidget.content.columnsMetadata}
                    sortByColumn={tableWidget.content.sortByColumn}
                    sortByDesc={tableWidget.content.sortByDesc}
                    significantDigitLabels={tableWidget.content.significantDigitLabels}
                    displayWithPages={tableWidget.content.displayWithPages}
                    showMobilePreview={showMobilePreview}
                />
            );
        case WidgetType.Metrics:
            const metricsWidget = widget as MetricsWidget;
            const metricsId = `metrics-${Utils.getShorterId(widget.id)}`;
            return (
                <MetricsWidgetComponent
                    id={metricsId}
                    title={!hideTitle && metricsWidget.showTitle ? metricsWidget.content.title : ""}
                    metrics={json}
                    metricPerRow={
                        metricsWidget.content.oneMetricPerRow ||
                        window.innerWidth < 640 ||
                        showMobilePreview
                            ? 1
                            : 3
                    }
                    significantDigitLabels={metricsWidget.content.significantDigitLabels}
                    metricsCenterAlign={metricsWidget.content.metricsCenterAlign}
                />
            );
        default:
            return null;
    }
}

export default WidgetRender;
