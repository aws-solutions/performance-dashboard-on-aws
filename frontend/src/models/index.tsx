/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
