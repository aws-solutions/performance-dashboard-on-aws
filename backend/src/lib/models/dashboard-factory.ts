import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
