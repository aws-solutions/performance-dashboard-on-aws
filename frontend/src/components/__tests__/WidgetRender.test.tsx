/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import WidgetRender from "../WidgetRender";
import { ChartWidget, ImageWidget, Widget, WidgetType } from "../../models";

test("renders a text widget", async () => {
    const textWidget = {
        widgetType: WidgetType.Text,
        content: { text: "Text widget" },
    } as Widget;
    const wrapper = render(<WidgetRender widget={textWidget} />);
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a chart widget", async () => {
    const chartWidget = {
        widgetType: WidgetType.Chart,
        content: { s3Key: { json: "data.json" } },
    } as ChartWidget;
    const wrapper = render(<WidgetRender widget={chartWidget} />);
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a table widget", async () => {
    const tableWidget = {
        widgetType: WidgetType.Table,
        content: { s3Key: { json: "data.json" } },
    } as Widget;
    const wrapper = render(<WidgetRender widget={tableWidget} />);
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a metrics widget", async () => {
    const metricsWidget = {
        widgetType: WidgetType.Metrics,
        content: { s3Key: { json: "data.json" } },
    } as Widget;
    const wrapper = render(<WidgetRender widget={metricsWidget} />);
    expect(wrapper.container).toMatchSnapshot();
});

test("renders an image widget", async () => {
    URL.createObjectURL = jest.fn();
    const textWidget = {
        widgetType: WidgetType.Image,
        content: { s3Key: { raw: "rawData" } },
    } as ImageWidget;
    const wrapper = render(<WidgetRender widget={textWidget} />);
    expect(URL.createObjectURL).toBeCalled();
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a section widget", async () => {
    const sectionWidget = {
        widgetType: WidgetType.Section,
        content: { s3Key: { json: "data.json" } },
    } as Widget;
    const wrapper = render(<WidgetRender widget={sectionWidget} />);
    expect(wrapper.container).toMatchSnapshot();
});
