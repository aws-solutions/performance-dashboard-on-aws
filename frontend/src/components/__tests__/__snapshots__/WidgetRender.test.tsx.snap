// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a chart widget 1`] = `
<div>
  <div
    class="widgetContainer"
  />
</div>
`;

exports[`renders a metrics widget 1`] = `
<div>
  <div
    class="widgetContainer"
  >
    <div
      aria-label=""
      class="padding-top-2"
      tabindex="-1"
    >
      
      <div>
        <div
          class="grid-col"
        >
          <div
            class="grid-row flex-column"
          >
            <div
              class="grid-col grid-row flex-row"
            >
              <div
                class="grid-col-4 padding-05"
              >
                <div
                  class="display-flex flex-column border-base-lightest border-2px height-card padding-1 overflow-x-auto overflow-y-auto"
                >
                  <div
                    class="flex-5"
                  >
                    <h3
                      class="text-base-darker text-bold margin-0 text-no-wrap"
                    />
                  </div>
                  <div
                    class="flex-3 usa-tooltip"
                    data-position="bottom"
                    title=""
                  >
                    <span
                      class="margin-0 text-no-wrap font-sans-xl text-bold"
                      style="color: rgb(41, 180, 187);"
                    />
                  </div>
                  <div
                    class="flex-2"
                  />
                  <div
                    class="flex-2"
                  />
                </div>
              </div>
              <div
                class="grid-col-4 padding-05"
              >
                <div
                  class="display-flex flex-column border-base-lightest border-2px height-card padding-1 overflow-x-auto overflow-y-auto"
                >
                  <div
                    class="flex-5"
                  >
                    <h3
                      class="text-base-darker text-bold margin-0 text-no-wrap"
                    />
                  </div>
                  <div
                    class="flex-3 usa-tooltip"
                    data-position="bottom"
                    title=""
                  >
                    <span
                      class="margin-0 text-no-wrap font-sans-xl text-bold"
                      style="color: rgb(41, 180, 187);"
                    />
                  </div>
                  <div
                    class="flex-2"
                  />
                  <div
                    class="flex-2"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a section widget 1`] = `
<div>
  <div
    class="widgetContainer"
  >
    <div>
      <div>
        
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a table widget 1`] = `
<div>
  <div
    class="widgetContainer"
  >
    <div
      aria-label=""
      class="overflow-x-hidden overflow-y-hidden padding-top-2"
      tabindex="-1"
    >
      
      <div
        class="Markdown usa-prose margin-top-0 margin-bottom-3 tableSummaryAbove textOrSummary"
      />
      <div
        class="overflow-x-hidden overflow-y-hidden"
      >
        <table
          class="usa-table usa-table--borderless"
          id="table-undefined"
          role="table"
        >
          <thead>
            <tr
              role="row"
            >
              <th
                aria-sort="none"
                colspan="1"
                role="columnheader"
                scope="col"
                style="min-width: 150px;"
              >
                <button
                  aria-label="Toggle SortBy: City"
                  class="usa-button usa-button--unstyled text-no-underline"
                  style="cursor: pointer;"
                  title="Toggle SortBy: City"
                  type="button"
                >
                  <span>
                    City
                  </span>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                    data-icon="chevron-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </th>
              <th
                aria-sort="none"
                colspan="1"
                role="columnheader"
                scope="col"
                style="min-width: 150px;"
              >
                <button
                  aria-label="Toggle SortBy: Jobs"
                  class="usa-button usa-button--unstyled text-no-underline"
                  style="cursor: pointer;"
                  title="Toggle SortBy: Jobs"
                  type="button"
                >
                  <span>
                    Jobs
                  </span>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                    data-icon="chevron-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            role="rowgroup"
          >
            <tr
              role="row"
            >
              <th
                role="cell"
                scope="row"
              >
                Seattle
              </th>
              <td
                role="cell"
              >
                300
              </td>
            </tr>
            <tr
              role="row"
            >
              <th
                role="cell"
                scope="row"
              >
                San Francisco
              </th>
              <td
                role="cell"
              >
                150
              </td>
            </tr>
          </tbody>
        </table>
        
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a text widget 1`] = `
<div>
  <div
    class="widgetContainer"
  >
    <div
      class=""
      id="text-undefined"
      tabindex="-1"
    >
      <div
        class="Markdown usa-prose textOrSummary"
      >
        <p>
          Text widget
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders an image widget 1`] = `
<div>
  <div
    class="widgetContainer"
  >
    <div
      aria-label=""
      class="preview-container padding-top-2"
      tabindex="-1"
    >
      
      <div
        class="Markdown usa-prose margin-top-0 margin-bottom-3 imageSummaryAbove textOrSummary"
      />
      <div>
        <img
          height="auto"
          width="auto"
        />
      </div>
    </div>
  </div>
</div>
`;
